File: linkedlist/876_MiddleOfLL.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head.next: return head
        curnode, length = head, 0
        while curnode:
            length, curnode = length+1, curnode.next
        curnode, curlength = head, 0
        while curnode:
            curlength, curnode = curlength+1, curnode.next
            if curlength == length // 2: return curnode

class SlowFastPointerSolution:
    def middleNode(self, head):
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

solution = Solution()
print(solution.middleNode(ListNode(val=1, next=ListNode(val=2, next=ListNode(val=3, next=ListNode(val=4, next=ListNode(val=5, next=ListNode(val=6))))))))

File: linkedlist/876_MiddleOfLL.cc
#include <unordered_map>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int i = 0;
        unordered_map<int, ListNode*> nodeMap;
        ListNode* temp = head;
        while(temp != nullptr){
            nodeMap[i] = temp;
            temp = temp->next;
            ++i;
        }

        int mid = i/2;
        return nodeMap[mid];
    }
};